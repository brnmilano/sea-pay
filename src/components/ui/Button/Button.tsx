import { ButtonSizes, ButtonVariants, type ButtonProps } from "./Button.type";
import styles from "./button.module.scss";

export function Button(props: ButtonProps) {
  const {
    text,
    buttonVariant = ButtonVariants.PRIMARY,
    size = ButtonSizes.SMALL,
    disabled = false,
    ariaLabel,
    ...rest
  } = props;

  return (
    <button
      className={`${styles.button} ${styles[buttonVariant]} ${
        styles[`size-${size}`]
      }`}
      aria-label={ariaLabel}
      data-variant={buttonVariant}
      data-size={size}
      disabled={disabled}
      {...rest}
    >
      {text}
    </button>
  );
}
