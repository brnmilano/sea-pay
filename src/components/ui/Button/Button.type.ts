import type { ButtonHTMLAttributes } from "react";

/**
 * Variantes visuais disponíveis para o componente Button
 * Define os estilos primário, secundário e de texto
 */
export enum ButtonVariants {
  PRIMARY = "primary",
  SECONDARY = "secondary",
  TEXT = "text",
}

/**
 * Tamanhos disponíveis para o componente Button
 * Define os tamanhos pequenos, médios e grandes
 */
export enum ButtonSizes {
  SMALL = "small",
  MEDIUM = "medium",
  LARGE = "large",
}

/**
 * Props para o componente Button
 * Extende as propriedades nativas do elemento button HTML
 */
export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  /** Texto a ser exibido no botão */
  text?: string;

  /** Variante visual do botão */
  buttonVariant?: ButtonVariants;

  /** Tamanho do botão */
  size?: ButtonSizes;

  /** Label para acessibilidade (aria-label) */
  ariaLabel?: string;

  /** Se o botão está desabilitado */
  disabled?: boolean;

  /** Referência para o elemento button */
  ref?: React.Ref<HTMLButtonElement>;
}
