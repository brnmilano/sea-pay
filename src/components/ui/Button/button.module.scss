.button {
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 0.5rem;
  border-radius: 0.375rem;
  border: none;
  font-size: 0.75rem;
  font-family: var(--font-family);
  height: 2.1875rem;
  cursor: pointer;
  transition:
    background-color 0.2s ease-in-out,
    border-color 0.2s ease-in-out,
    transform 0.1s ease-in-out;
  font-weight: 700;
  outline: none;

  &:focus-visible {
    outline: 2px solid var(--brand-color-500);
    outline-offset: 2px;
  }

  &:active {
    transform: scale(0.98);
  }

  &:disabled {
    cursor: not-allowed;
    opacity: 0.6;
    pointer-events: none;
  }

  &.primary {
    background-color: var(--brand-color-500);
    color: var(--semantic-white);
    width: 100%;

    &:hover:not(:disabled) {
      background-color: var(--brand-color-300);
    }
  }

  &.secondary {
    background-color: var(--semantic-white);
    color: var(--brand-color-500);
  }

  &.text {
    background-color: transparent;
    color: var(--brand-color-500);
    font-weight: 700;
    border: 1px solid transparent;

    &:hover:not(:disabled) {
      border-color: var(--brand-color-500);
      background-color: var(--brand-color-100);
    }
  }

  &.size-small {
    width: 100%;
    max-width: 4.625rem;
    padding: 0 0.75rem;
  }

  &.size-medium {
    width: 100%;
    max-width: 6.875rem;
    padding: 0 1rem;
  }

  &.size-large {
    width: 100%;
    max-width: 24.0625rem;
    padding: 0 1.5rem;
  }
}
