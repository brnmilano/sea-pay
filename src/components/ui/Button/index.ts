// Barrel export para o componente Button
export { Button } from "./Button";
export type { ButtonProps } from "./Button.type";
