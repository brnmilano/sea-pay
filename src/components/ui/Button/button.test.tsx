import { render, screen } from "@testing-library/react";
import { Button } from "./Button";
import { ButtonSizes, ButtonVariants } from "./Button.type";

describe("Button Component", () => {
  test("should match snapshot.", () => {
    const { container } = render(
      <Button
        text="Snapshot Button"
        ariaLabel="Snapshot Button"
        size={ButtonSizes.MEDIUM}
        buttonVariant={ButtonVariants.PRIMARY}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test("should check if the passed ref points to the correct DOM element.", () => {
    const ref = { current: null };

    render(
      <Button
        text="Ref Button"
        ariaLabel="Ref Button"
        size={ButtonSizes.MEDIUM}
        buttonVariant={ButtonVariants.PRIMARY}
        ref={ref}
      />,
    );

    const button = screen.getByRole("button", { name: /ref button/i });

    expect(button).toBeInTheDocument();
    expect(ref.current).toBe(button);
  });

  test("should render the button with correct text and aria-label.", () => {
    render(
      <Button
        text="Click Me"
        ariaLabel="Click Me"
        size={ButtonSizes.MEDIUM}
        buttonVariant={ButtonVariants.PRIMARY}
      />,
    );

    const button = screen.getByRole("button", { name: /click me/i });

    expect(button).toBeInTheDocument();
    expect(button).toHaveTextContent("Click Me");
  });

  test("should call onClick handler when clicked.", () => {
    const handleClick = jest.fn();

    render(
      <Button
        text="Click Me"
        ariaLabel="Click Me"
        size={ButtonSizes.MEDIUM}
        buttonVariant={ButtonVariants.PRIMARY}
        onClick={handleClick}
      />,
    );

    const button = screen.getByRole("button", { name: /click me/i });

    expect(button).toBeInTheDocument();

    button.click();
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  test("should check keyboard focus (tab) and visible focus indicator.", () => {
    render(
      <Button
        text="Focus Me"
        ariaLabel="Focus Me"
        size={ButtonSizes.MEDIUM}
        buttonVariant={ButtonVariants.PRIMARY}
      />,
    );

    const button = screen.getByRole("button", { name: /focus me/i });

    expect(button).toBeInTheDocument();
    expect(document.body).toHaveFocus();

    button.focus();
    expect(button).toHaveFocus();
  });

  test("should check reading by screen reader (aria-label, aria-pressed, etc.)", () => {
    render(
      <Button
        text="Aria Button"
        ariaLabel="Aria Button"
        size={ButtonSizes.MEDIUM}
        buttonVariant={ButtonVariants.PRIMARY}
        aria-pressed="false"
      />,
    );

    const button = screen.getByRole("button", { name: /aria button/i });

    expect(button).toBeInTheDocument();
    expect(button).toHaveAttribute("aria-label", "Aria Button");
    expect(button).toHaveAttribute("aria-pressed", "false");
  });

  test("should if the onClick event is fired when clicked.", () => {
    const handleClick = jest.fn();

    render(
      <Button
        text="Clickable Button"
        ariaLabel="Clickable Button"
        size={ButtonSizes.MEDIUM}
        buttonVariant={ButtonVariants.PRIMARY}
        onClick={handleClick}
      />,
    );

    const button = screen.getByRole("button", { name: /clickable button/i });

    expect(button).toBeInTheDocument();

    button.click();
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  test("should check if it does not fire onClick when disabled.", () => {
    const handleClick = jest.fn();

    render(
      <Button
        text="Disabled Button"
        ariaLabel="Disabled Button"
        size={ButtonSizes.MEDIUM}
        buttonVariant={ButtonVariants.PRIMARY}
        onClick={handleClick}
        disabled
      />,
    );

    const button = screen.getByRole("button", { name: /disabled button/i });

    expect(button).toBeInTheDocument();
    expect(button).toBeDisabled();

    button.click();
    expect(handleClick).not.toHaveBeenCalled();
  });

  test("should verify that form attributes work as expected when used in forms.", () => {
    render(
      <form>
        <Button
          text="Submit Button"
          ariaLabel="Submit Button"
          size={ButtonSizes.MEDIUM}
          buttonVariant={ButtonVariants.PRIMARY}
          type="submit"
        />
      </form>,
    );

    const button = screen.getByRole("button", { name: /submit button/i });

    expect(button).toBeInTheDocument();
    expect(button).toHaveAttribute("type", "submit");
  });

  test("should apply SMALL size and PRIMARY variant.", () => {
    const { container } = render(
      <Button
        text="Test Button"
        ariaLabel="Test Button"
        size={ButtonSizes.SMALL}
        buttonVariant={ButtonVariants.PRIMARY}
      />,
    );

    const button = screen.getByRole("button", { name: /test button/i });

    expect(button).toBeInTheDocument();
    expect(button).toHaveTextContent("Test Button");
    expect(container.firstChild).toHaveClass("size-small");
    expect(container.firstChild).toHaveClass("primary");
  });

  test("should apply MEDIUM size and SECONDARY variant.", () => {
    const { container } = render(
      <Button
        text="Test Button"
        ariaLabel="Test Button"
        size={ButtonSizes.MEDIUM}
        buttonVariant={ButtonVariants.SECONDARY}
      />,
    );

    const button = screen.getByRole("button", { name: /test button/i });

    expect(button).toBeInTheDocument();
    expect(button).toHaveTextContent("Test Button");
    expect(container.firstChild).toHaveClass("size-medium");
    expect(container.firstChild).toHaveClass("secondary");
  });

  test("should apply LARGE size and TEXT variant.", () => {
    const { container } = render(
      <Button
        text="Test Button"
        ariaLabel="Test Button"
        size={ButtonSizes.LARGE}
        buttonVariant={ButtonVariants.TEXT}
      />,
    );

    const button = screen.getByRole("button", { name: /test button/i });

    expect(button).toBeInTheDocument();
    expect(button).toHaveTextContent("Test Button");
    expect(container.firstChild).toHaveClass("size-large");
    expect(container.firstChild).toHaveClass("text");
  });
});
