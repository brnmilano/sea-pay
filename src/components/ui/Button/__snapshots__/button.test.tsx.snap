// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Button Component should match snapshot. 1`] = `
<div>
  <button
    aria-label="Snapshot Button"
    class="button primary size-medium"
    data-size="medium"
    data-variant="primary"
  >
    Snapshot Button
  </button>
</div>
`;
